docs/llama_multi_lora_tutorial.md

```diff
@@ -0,0 +1,243 @@
+
+> Before you continue reading, it's important to note that all command-line instructions containing `<xx.yy>` in the document cannot be used directly by copying and pasting. 
+>
+> `<xx.yy>` represents the Triton version, and you must specify the Triton version you want to use for the bash command to work.
+
+## Step 1: Start a docker container for triton-vllm serving
+
+**A docker container is strongly recommended for serving**, and this tutorial will only demonstrate how to launch triton in docker env.
+
+First, create a docker container using the NCG built for vllm serving:
+
+```bash
+# NOTICE: you must first cd to your vllm_workspace path outside the container.
+mkdir vllm_workspace && cd vllm_workspace
+
+sudo docker run --gpus all -it --net=host -p 8001:8001 --shm-size=12G \
+--ulimit memlock=-1 --ulimit stack=67108864 -v ${PWD}:/vllm_workspace \
+-w /vllm_workspace nvcr.io/nvidia/tritonserver:<xx.yy>-vllm-python-py3 \
+/bin/bash
+```
+
+**NOTICE:** the version of triton docker image should be configurated, here we use `<xx.yy>` to symbolize.
+
+Triton's vLLM container has been introduced starting from 23.10 release, and `multi-lora` experimental support was added in vLLM v0.3.0 release.
+
+> Docker image version `nvcr.io/nvidia/tritonserver:23.12-vllm-python-py3` or higher version is strongly recommended.
+
+---
+
+For **pre-24.03 containers**, the docker images didn't support multi-lora feature, so you need to replace that provided in the container `/opt/tritonserver/backends/vllm/model.py` with the most up to date version. Just follow this command:
+
+Download the `model.py` script from github:
+
+```bash
+wget -P /opt/tritonserver/backends/vllm/ https://raw.githubusercontent.com/triton-inference-server/vllm_backend/r<xx.yy>/src/model.py
+```
+
+**Notice:** `r<xx.yy>` is the triton version you need to configure.
+
+This command will download the `model.py` script to the Triton vllm backend directory to support multi-lora feature.
+
+## Step 2: Install vLLM with multi-lora feature
+
+We are now in the docker container, and **the following operations will be done in container environment.**
+
+```bash
+cd /vllm_workspace
+```
+
+**NOTICE**: To enable multi-lora feature and speed up the inference, developers have integrated punica kernels into the `csrc` directory. To compile the punica kernels, you need to turn the `VLLM_INSTALL_PUNICA_KERNELS` env variable on to allow punica kernels compilation.
+
+By default, the punica kernels will **NOT** be compiled when installing the vLLM.
+
+__2.1 install with pip__
+
+For Triton version before 24.02, you need the following command:
+
+```bash
+VLLM_INSTALL_PUNICA_KERNELS=1 pip install vllm==0.3.0
+```
+
+__2.2 build from source (Recommended)__
```

Comment:
Why is this option recommended?

Reply:
Sorry, it is a historical issue, when multi-lora feature hasn't been added to vLLM in early time, developers can only install vllm through source code to enable this feature. 

Now this tip can be removed :)